music/src/gansynth/model.ts

```diff
@@ -0,0 +1,292 @@
+/**
+ * Core implementation for [GANSynth]{@link
+ * https://g.co/magenta/gansynth} models.
+ *
+ * @license
+ * Copyright 2019 Google Inc. All Rights Reserved.
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Imports.
+ */
+import * as tf from '@tensorflow/tfjs';
+
+import * as logging from '../core/logging';
+
+import {MIDI_PITCHES, MIN_MIDI_PITCH, N_LATENTS, N_PITCHES} from './constants';
+import {boxUpscale, initialPad, pixelNorm} from './custom_layers';
+
+/**
+ * Main "GANSynth" neural synthesizer model class.
+ */
+class GANSynth {
+  private checkpointURL: string;
+  private initialized: boolean;
+  private outputShape: number[];
+  private readonly nn = tf.sequential();
+
+  /**
+   * `GANSynth` constructor.
+   *
+   * @param checkpointURL Path to the checkpoint directory.
+   * @param chunkLength The length of chunks (excluding receptive field
+   * padding). Sequences longer than this amount will be split into batches of
+   * this size for processing.
+   */
+  constructor(checkpointURL: string) {
+    this.checkpointURL = checkpointURL;
+  }
+
+  /**
+   * Loads variables from the checkpoint and builds the model graph.
+   */
+  async initialize() {
+    this.dispose();
+    const startTime = performance.now();
+    const vars = await fetch(`${this.checkpointURL}/weights_manifest.json`)
+                     .then((response) => response.json())
+                     .then(
+                         (manifest: tf.io.WeightsManifestConfig) =>
+                             tf.io.loadWeights(manifest, this.checkpointURL));
+
+    // Rescale for He initialization
+    // Training initialized to N(0,1) and then rescaled output, so here we get
+    // the same result by rescaling the saved weights.
+    for (const v in vars) {
+      if (v.includes('kernel')) {
+        const fanIn = vars[v].shape[0] * vars[v].shape[1] * vars[v].shape[2];
+        vars[v] = tf.mul(vars[v], tf.sqrt(2 / fanIn));
```

Comment:
I think these are leaking memory: The intermediate values of `tf.sqrt` and the overwritten value of `vars[v]` are never disposed.